src/codec.rs

```diff
@@ -286,6 +284,52 @@ pub trait DecodeLength {
 	fn len(self_encoded: &[u8]) -> Result<usize, Error>;
 }
 
+/// A decoding context.
+///
+/// Used in [`Decode::decode_into`] to allow the implementation to explicitly
+/// assert that the `MaybeUninit` passed into that function was properly initialized.
+pub struct DecodeContext(PhantomData<*const ()>);
+
+/// A marker type signifying that the decoding finished.
+pub struct DecodeFinished(PhantomData<*const ()>);
```

Comment:
Also as discussed this could be moved somewhere else to avoid being able to just instantiate `DecodeFinished` in this file.

Reply:
1. The `PhantomData<*const ()>` makes the type `!Send` and `!Sync` because of the `*const ()` inside, which makes it harder to accidentally misuse. (Not strictly necessary, but nice to have.)
2. A type like `struct Foo` is publicly constructible anywhere, so e.g. if someone would implement the `Decode::decode_into` *outside* of this crate they'd be able to construct `DecodeFinished` without calling `assert_decoding_finished`.